Support `sortedTake` in beam runner

nownikhil opened the pull request:
We flatten all the input PriorityQueues and construct a new PQ using the monoid provided.
TESTS: Updated failing unit test

nownikhil committed:
Support `sortedTake` in beam runner

We flatten all the input PriorityQueues and construct a new PQ using the monoid provided.
TESTS: Updated failing unit test

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
   )(implicit ordK: Ordering[K], kryoCoder: KryoCoder): PCollection[KV[K, java.lang.Iterable[U]]] = {
     reduceFn match {
       case ComposedMapGroup(f, g) => planMapGroup(planMapGroup(pcoll, f), g)
+      case EmptyGuard(MapValueStream(SumAll(pqm: PriorityQueueMonoid[V]))) =>
```
I don't *think* this should either compile correctly. I probably wrote some bad code in my example. The `SumAll` type is `TraversableOnce[A] => Iterator[A]` where in this case `A` = `V`. So when you pattern match on `PriorityQueueMonoid[T]` that means `V = PriorityQueue[T]`, so, it shouldn't be `PriorityQueueMonoid[V]`. A little known fact is that you can use a lower case to bind an unknown type: `pqm: PriorityQueueMonoid[v]` and then use `v` lower down.

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
+              // We are not using plus method defined in PriorityQueueMonoid as it is mutating
+              // input Priority Queues. We create a new PQ from the individual ones.
+              // We didn't use Top PTransformation in beam as it is not needed, also
+              // we cannot access `max` defined in PQ monoid.
```
why not extract the comparator: `val cmp = pqm.zero.comparator()` and then use Top.of as @tlazaro suggested here:

https://github.com/twitter/scalding/issues/1947#issuecomment-923522362

nownikhil replied:
Top requires `max` no of elements in its constructor. But we cannot access it from the monoid.

johnynek replied:
ahh... man that PriorityQueueMonoid is really bad. :(

ahh, but there is a way out!

https://github.com/twitter/algebird/blob/0c45f9395020ceecd28899fe156b7e3c92b29814/algebird-core/src/main/scala/com/twitter/algebird/mutable/PriorityQueueMonoid.scala#L27

you could subclass it! `class ScaldingPriorityQueueMonoid[K](val count: Int)(implicit val ordering: Ordering[K]) extends PriorityQueueMonoid(count)(ordering)` and then you can access the count and the ordering.

johnynek replied:
I made this quick PR: https://github.com/twitter/algebird/pull/1008

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
+              // input Priority Queues. We create a new PQ from the individual ones.
+              // We didn't use Top PTransformation in beam as it is not needed, also
+              // we cannot access `max` defined in PQ monoid.
+              val flattenedValues = input.getValue.asScala.flatMap { value =>
```
I think this is going to materialize everything in memory. I think you should do `input.getValue.asScala.toStream.flatMap`... to make sure this is a lazy Iterable.

nownikhil replied:
`input.getValue.asScala` gives an Iterator only and `flatmap` also works lazily. Which call would materialize it?

johnynek replied:
`getValue` returns the `Iterable[_]` no? so, asScala on that is going to give you an `Iterable` if I'm not mistaken. You could call `.iterator` on that before you call .asScala and that will certainly be lazy, which is better than the `toStream` since we would rather not materialize the whole set.

johnynek replied:
actually... `build` requires an `Iterable` and since `flattenedValues` is a `val` it will pin things in memory.

To take this suggestion, I think you need `@inline def flattenedValues = input.getValue.asScala.toStream.flatMap...`

tlazaro replied:
I think I understand all the pieces but would really appreciate if you could clarify how to make it properly lazy.

In the first part are you referring to `Iterable` keeping references to all elements, as opposed to `Iterator` which wouldn't?

Then when you mention using inline, is it to avoid a closure over that variable keeping the reference alive for longer or landing in more expensive generation? Or more about escape analysis or similar?

johnynek replied:
In scala, most Iterable methods (like flatMap) are not lazy and will materialize the entire result in many cases (basically the only exception is Stream or LazyList in scala 2.13).

An Iterator generally drops references to the items it has iterated past, although you could imagine making one that doesn't. Consider a List iterator:
```scala
class ListIterator[A](lst: List[A]) extends Iterator[A] {
  private[this] var current: List[A] = lst
  def hasNext: Boolean = current.nonEmpty
  def next(): A = {
    val result = current.head
    current = current.tail
    result
  }
}
```
The way the scala compiler works, `lst` is only referenced by the constructor, so you won't keep a reference to that beyond the constructor which just uses it to initialize.

Finally, the `@inline` is just a hint, but the `def` means there is nothing keeping the reference around so when the GC runs it could colllect anything not pointed to by the Stream at that moment.

BUT, all of this is a bit moot because looking inside `PriorityQueueMonoid` we see that `build` traverses the Iterable twice... so if it is a large `Stream` you will materialize the whole thing...

I think the best idea would be just subclass PriorityQueueMonoid so you can access the max size and the ordering, and use that call Top.of.

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
+        pcoll.apply(MapElements.via(
+          new SimpleFunction[KV[K, java.lang.Iterable[V]], KV[K, java.lang.Iterable[U]]]() {
+            override def apply(input: KV[K, lang.Iterable[V]]): KV[K, java.lang.Iterable[U]] = {
+              // We are not using plus method defined in PriorityQueueMonoid as it is mutating
```
I think you also need to match on:

https://github.com/twitter/scalding/blob/e1988110f61a2849a077b92250e1e1d58511cd90/scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala#L271

since I guess this restriction is also enforced on the mapside operations... 

You'll need a slightly different approach there (basically a custom cache).

But maybe they don't do their mutation detection on the mappers?

nownikhil replied:
Tried with a unit test and it failed because of mutation. Instead of overwriting the `put` method in the `SummingCache` I was hoping to pattern match on Semigroup and overwrite the plus method.

```
case class ImmutablePQMonoid[T](
    size: Int
  )(implicit ord: Ordering[T]) extends PriorityQueueMonoid[T](size) {
    override def plus(left: PriorityQueue[T], right: PriorityQueue[T]): PriorityQueue[T] = {
      super.build(left.iterator().asScala.toIterable ++ right.iterator().asScala.toIterable)
    }
  }
```
But the same issue is there. I cannot access the size field.

nownikhil replied:
Or maybe this?

```
  case class ImmutablePQMonoid[T](
    pqm: PriorityQueueMonoid[T]
  )(implicit ord: Ordering[T]) extends Monoid[PriorityQueue[T]] {
    override def zero: PriorityQueue[T] = pqm.zero

    override def plus(
      l: PriorityQueue[T],
      r: PriorityQueue[T]
    ): PriorityQueue[T] = ???
  }
```

johnynek replied:
why not just pattern match on this monoid and not use it at all? Like, as a short fix, you could just make sumByLocalKeys a no-op for this particular monoid in the short term, but for a better fix you could implement a summingcache for this particular monoid.

I think making a full copy of the priority queue each time will be a perf killer and I doubt it will be better than just not using a mapside cache at all...

The more we talk about this, the more it seems like maybe copying the immutable heap in from cats-collections is the right move (or send a PR to algebird and we can publish a new version and use it from there, either way...)

Seems like we keep getting paper cuts here.

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -102,23 +101,16 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
-          .groupAll
-          .sortedReverseTake(3),
-        Seq(5, 4, 3)
+  test("sortedTake"){
```
note that `bufferedTake` also uses the problematic monoid internally. Worth testing:

https://github.com/twitter/scalding/blob/e1988110f61a2849a077b92250e1e1d58511cd90/scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala#L662

johnynek commented:
Good lesson here: mutation, NOT EVEN ONCE!

nownikhil committed:
Added extension of PriorityQueueMonoid

Added ScaldingPriorityQueueMonoid which exposes count which we later use in TopCombineFn.
Added unit test for bufferedTake
Disabled map side aggregation when using ScaldingPriorityQueueMonoid

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -52,6 +57,10 @@ sealed abstract class BeamOp[+A] {
     parDo(FlatMapFn(f))
 }
 
+private final case class SerializableComparator[T](comp: Comparator[T]) extends Comparator[T] {
```
note, you may want to use 
https://github.com/twitter/scalding/blob/a9500fd154fa3a318e2fad063e6d3d15e70e8b17/scalding-core/src/main/scala/com/twitter/scalding/serialization/Externalizer.scala#L29

(and in other locations), e.g.
```scala
private class SerializableComparator[A](comp: Comparator[A]) extends Comparator[A] {
  private[this] val extCmp = Externalizer(comp)
  override def compare(o1: T, o2: T): Int = extCmp.get.compare(o1, o2)
}
```
The `Externalizer` tries both java serialization and kryo serialization. You may want to use it for some of the beamops as well with the functions (scalding on cascading does use Externalizer).

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -61,19 +70,24 @@ object BeamOp extends Serializable {
+                input.getValue.asScala.toStream.flatMap(_.asScala.toStream)
+
+              val outputs: java.util.List[v] = topCombineFn.apply(flattenedValues.asJava)
+              val pqs = pqm.build(outputs.asScala)
```
why do we use both `TopCombineFn` and `pqm`? I'm confused why one isn't enough? Seems like we should be able to return `output` to me.

nownikhil replied:
`U` here is `PriorityQueue[v]` so casting to `U` fails without converting the `outputs` to a PriorityQueue.

johnynek replied:
Ahh. Okay. It makes sense now. Can you comment?

johnynek commented on scalding-beam/src/test/scala/com/twitter/scalding/beam_backend/BeamBackendTests.scala:
```diff
@@ -113,6 +115,18 @@ class BeamBackendTests extends FunSuite with BeforeAndAfter {
+      TypedPipe
+        .from(1 to 50)
+        .groupAll
+        .bufferedTake(100)
```
just a note: this is going to be really bad in a real job without map-side aggregation: the key is `Unit` so there is only one key, so this would have each mapper send 100, then have the reducers pick 100 of those.

But with no mapside aggregation, all the data will be sent to the reducers, and they will throw away all but 100.

But we can add an issue and come back and address this.

nownikhil replied:
Opened a ticket for this.
https://github.com/twitter/scalding/issues/1952

johnynek commented on scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala:
```diff
@@ -59,6 +70,30 @@ object BeamOp extends Serializable {
+          new SimpleFunction[KV[K, java.lang.Iterable[PriorityQueue[v]]], KV[K, java.lang.Iterable[U]]]() {
+            override def apply(input: KV[K, lang.Iterable[PriorityQueue[v]]]): KV[K, java.lang.Iterable[U]] = {
+
+              val topCombineFn = new TopCombineFn[v, SerializableComparator[v]](
```
I think you can lift this out of the `apply` method to reduce allocations. We shouldn't allocate a new one on every input should we?

nownikhil committed:
Use externalizer in SerializableComparator

tlazaro commented:
I'm fine with landing this, how do we usually do it? 'Squash and merge' ?

johnynek commented:
yes, squash and merge if you think it is ready. Looks good to me too.
